internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,289 @@
+				Name:     "kargo-controller-secrets-readonly",
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind:      "ServiceAccount",
+					Name:      sa.Name,
+					Namespace: r.cfg.KargoNamespace,
+				},
+			},
+		}
+
+		// Create a new RoleBinding if it doesn't exist, or update it if it does.
+		if err := r.client.Get(ctx, client.ObjectKey{Name: roleBindingName, Namespace: project.Name}, roleBinding); err != nil {
+			if kubeerr.IsNotFound(err) {
+				// Create RoleBinding
+				if err := r.createRoleBindingFn(ctx, roleBinding); err != nil {
+					return fmt.Errorf("error creating RoleBinding %q for ServiceAccount %q in Project namespace %q: %w",
+						roleBinding.Name, sa.Name, project.Name, err)
+				}
+				logger.Info("Created RoleBinding for ServiceAccount", "roleBinding", roleBindingName, "namespace", project.Name)
+			} else {
+				return fmt.Errorf("error retrieving RoleBinding %q in namespace %q: %w", roleBindingName, project.Name, err)
+			}
+		} else {
+			// Update RoleBinding
+			if err := r.client.Update(ctx, roleBinding); err != nil {
+				return fmt.Errorf("error updating RoleBinding %q in Project namespace %q: %w", roleBinding.Name, project.Name, err)
+			}
+			logger.Debug("Updated RoleBinding for ServiceAccount", "roleBinding", roleBindingName, "namespace", project.Name)
+		}
+	}
+	logger.Info("Completed creating RoleBindings for all Projects")
+	return nil
+}
+
+// removeControllerPermissions removes RoleBindings for the specified ServiceAccount
+// from all Projects. This function is called when the ServiceAccount is no longer
+// managed by the controller, ensuring that permissions are cleaned up to prevent
+// unauthorized access to resources.
+func (r *reconciler) removeControllerPermissions(ctx context.Context, sa *corev1.ServiceAccount) error {
+
+	roleBindingName := fmt.Sprintf("%s-readonly-secrets", sa.Name)
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceaccount", sa.Name,
+		"roleBinding", roleBindingName,
+	)
+	logger.Info("Starting to delete RoleBindings for all Projects")
+
+	projectList := &kargoapi.ProjectList{}
+	if err := r.client.List(ctx, projectList); err != nil {
+		return fmt.Errorf("error listing Projects: %w", err)
+	}
+
+	for _, project := range projectList.Items {
+
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: project.Namespace,
+			},
+		}
+
+		if err := r.deleteRoleBindingFn(ctx, roleBinding); err != nil {
+			if kubeerr.IsNotFound(err) {
+				logger.Debug("RoleBinding not found, skipping deletion", "roleBinding", roleBindingName, "namespace", project.Namespace)
+				continue // Skip to the next project if RoleBinding is not found
+			}
+
+			// Log the error but continue to the next project
+			logger.Error(err, "Failed to delete RoleBinding", "roleBinding", roleBindingName, "namespace", project.Namespace)
+			continue // Continue with the next project even after an error
```

Comment:
This isn't the right thing to do. This can lead to apparent success in cleaning up these RBs when cleanup is, in fact, not complete.

Return the error instead.

When the error is returned from here, the finalizer should not be removed, and the SA should be re-queued for another reconciliation attempt.